File: src/main/java/ludwig/ide/CodeTreeView.java
package ludwig.ide;

import com.sun.javafx.scene.control.skin.TreeViewSkin;
import javafx.scene.control.TreeItem;
import javafx.scene.control.TreeView;
import ludwig.model.*;
import ludwig.workspace.Environment;

import java.util.List;


public class CodeTreeView extends TreeView<Node> {
    private final Environment environment;

    public CodeTreeView(Environment environment) {
        super(new TreeItem<>(new ListNode()));
        this.environment = environment;
        setShowRoot(false);
    }

    public void setNode(Node<?> node) {
        getRoot().getChildren().clear();

        if (node instanceof FunctionNode) {
            for (int i = 0; i < node.children().size(); i++) {
                if (!(node.children().get(i) instanceof  VariableNode)) {
                    setContent(getRoot(), node.children().subList(i, node.children().size()));
                    break;
                }
            }
        } else if (node instanceof OverrideNode) {
            for (int i = 1; i < node.children().size(); i++) {
                if (!(node.children().get(i) instanceof  VariableNode)) {
                    setContent(getRoot(), node.children().subList(i, node.children().size()));
                    break;
                }
            }
        }
    }

    private void setContent(TreeItem<Node> parent, List<Node<?>> nodes) {
        for (Node<?> node: nodes) {
            TreeItem<Node> item = new TreeItem<>(node);
            parent.getChildren().add(item);
            item.setExpanded(true);
            setContent(item, node.children());
        }
    }

    public Node<?> selectedNode() {
        TreeItem<Node> selectedItem = getSelectionModel().getSelectedItem();
        return selectedItem != null ? selectedItem.getValue() : null;
    }

    public void locate(Node<?> node) {

    }
}

File: src/main/java/ludwig/ide/CodeTreeView.kt
package ludwig.ide

import com.sun.javafx.scene.control.skin.TreeViewSkin
import javafx.scene.control.TreeItem
import javafx.scene.control.TreeView
import ludwig.model.*
import ludwig.workspace.Environment


class CodeTreeView(private val environment: Environment) : TreeView<Node<*>>(TreeItem<Node<*>>(ListNode())) {

    init {
        isShowRoot = false
    }

    fun setNode(node: Node<*>?) {
        root.children.clear()

        if (node is FunctionNode) {
            for (i in 0..node.children().size - 1) {
                if (node.children()[i] !is VariableNode) {
                    setContent(root, node.children().subList(i, node.children().size))
                    break
                }
            }
        } else if (node is OverrideNode) {
            for (i in 1..node.children().size - 1) {
                if (node.children()[i] !is VariableNode) {
                    setContent(root, node.children().subList(i, node.children().size))
                    break
                }
            }
        }
    }

    private fun setContent(parent: TreeItem<Node<*>>, nodes: List<Node<*>>) {
        for (node in nodes) {
            val item = TreeItem(node)
            parent.children.add(item)
            item.isExpanded = true
            setContent(item, node.children())
        }
    }

    fun selectedNode(): Node<*>? {
        val selectedItem = selectionModel.selectedItem
        return selectedItem?.value
    }

    fun locate(node: Node<*>) {

    }
}
